### PR TITLE
ScalametaParser: allow `erased` keyword in types

### LINKED ISSUE #2774: Support erased definitions
This template is mostly a guideline, not a strict requirement, except
the sections marked `(required)`.

Please make sure that the issue is observable using the latest version of `scalafmt` and
within the `scalafmt` **CLI**: `https://scalameta.org/scalafmt/docs/installation.html#cli`

Using the CLI is _mandatory_, in one of these configurations:
- latest version of CLI and latest version in `.scalafmt.conf`
- previous version of CLI and latest version in `.scalafmt.conf`
- latest version of CLI and previous version in `.scalafmt.conf`

## Configuration (required) ##

Please paste the contents of your `.scalafmt.conf` file here:
```
version = 3.5.8
binPack.parentConstructors = true
maxColumn = 120
includeCurlyBraceInSelectChains = true
align.preset = most
trailingCommas = preserve
runner.dialect = scala3
project.excludeFilters = ["kaivessel/package.scala"]

rewrite.scala3.insertEndMarkerMinLines = 5
...
```

NB: before submitting, please confirm that the problem is observed in the
*latest published* version of the formatter! We don't publish hotfixes for older
versions, and the problem you have observed in an older version may have already
been fixed.

## Command-line parameters (required) ##

When I run scalafmt via CLI like this: `<command-line parameter>`

If you were not using CLI initially:
- please check using `scalafmt` CLI before submitting here (required)
- if the problem does not present via CLI, DO NOT SUBMIT
  here: this is NOT a formatter issue
  - it could be an issue with your integration (sbt-scalafmt, metals, intellij etc.)
  - instead please submit the issue in the appropriate tool's repository
  - for instance, if the tool is `sbt` with the `sbt-scalafmt` plugin: please
    submit in `sbt-scalafmt` and include scalafmt parameters from `build.sbt`
    and the `sbt` command used
  - conversely, do not submit formatting issues into other tools; you must first
    verify using `scalafmt` CLI

## Steps

Given code like this:
```scala
package kuraga.datum

import scala.language.experimental.erasedDefinitions

import Repr.Holder

object Repr:
  trait Holder:
    type A

trait Repr[-P[-_, +_], +X]:
  def apply(erased holder: Holder)(p: P[X, holder.A]): holder.A
```

## Problem

Scalafmt formats raises an error like this:
```scala
Repr.scala:12: error: [dialect scala3] : expected but identifier found
  def apply(erased holder: Holder)(p: P[X, holder.A]): holder.A
```

## Workaround

I've found that by...

## Notes

See also...

### PR DESCRIPTION
**Pull request by kitbellew**
Fixes #2774.

### CODE REVIEW

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 1]
```diff
@@ -654,7 +654,7 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
       var hasTypes = false
 
       @tailrec
-      def paramOrType(): Type = token match {
+      def paramOrType()(implicit modsBuf: mutable.Builder[Mod, List[Mod]]): Type = token match {
```
Comment (tgodzik) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:657]: Is making it implicit necessary?

**Review by tgodzik — APPROVED** [review]
One minor comment, otherwise LGTM!

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 1 above)
Comment (kitbellew) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:657]: of course, not 😀 old habits, reducing size of diff.

Comment (kitbellew) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: done

**Pull request merged**